File: StateFB.m
%% System Ax + Bu
A     = [0 1 0 0; 0 0 1 0; 0 0 0 1; 0 0 0 0];
B     = [0; 0; 0; 1]; 
B1     = [0; 0; 0; 9.8];
p     = [-1,-2,-3,-4];
K     = place(A,B,p); %Controller gains

dt    = 0.001;  %Time step
t     = 50;     %Sim time
s     = 0:dt:t; %Time

x = zeros(size(A,1),length(s)); %State vector
x(:,1) = [0; 0; 0; 0];    %Initial condotions
e = zeros(size(A,1),length(s)); %Error vector
r = zeros(size(A,1),length(s)); %Reference vector
u = zeros(1,length(s));
%% Solver and control
for i = 1 : length(s)
    r(:,i) = [4 + sin(0.5*(i-1)*dt); 0.5*cos(0.5*(i-1)*dt); -0.25*sin(0.5*(i-1)*dt); -0.125*cos(0.5*(i-1)*dt)];%Reference
    xpppp  = 0.0625*sin(0.5*(i-1)*dt);
    e(:,i) = x(:,i) - r(:,i); %Error
    
    u(i) = xpppp + (-K(1)*e(1,i)-K(2)*e(2,i)-K(3)*e(3,i)-K(4)*e(4,i))/9.8;   %State FeedBack
    % u =  -K(1)*e(1,i)-K(2)*e(2,i)-K(3)*e(3,i)-K(4)*e(4,i);   %State FeedBack

    x(:,i+1) = x(:,i) + dt*(A*x(:,i) + B1*u(i)- [0;0;0;xpppp]); %Forward Euler
end

%% Plots
%Position plot
figure(1)
grid on; hold on;
plot(s,x(1,1:end-1), 'LineWidth', 2)
plot(s,r(1,:), 'LineWidth', 2);
xlabel({'Time(s)'}, 'interpreter', 'latex', 'fontsize', 14)
ylabel({'Position(m)'}, 'interpreter', 'latex', 'fontsize', 14)
legend({'$x_{1}(t)$','$x_{1ref}(t)$'}, 'interpreter', 'latex', 'location', 'best', 'fontsize', 14);


%Errors (e1->x, e2->xp, e3->xpp, e4->xppp)
figure(2)
grid on; hold on;
plot(s,e, 'LineWidth', 2);
xlabel({'Time'}, 'interpreter', 'latex', 'fontsize', 14)
ylabel({'Errors'}, 'interpreter', 'latex', 'fontsize', 14)
legend({'$e_{1}(t)$','$e_{2}(t)$','$e_{3}(t)$','$e_{4}(t)$'}, 'interpreter', 'latex', 'location', 'best', 'fontsize', 14);

%Control input plot
figure(3)
grid on; hold on;
plot(s,u, 'LineWidth', 2)
xlabel({'Time'}, 'interpreter', 'latex', 'fontsize', 14)
ylabel({'Control input'}, 'interpreter', 'latex', 'fontsize', 14)

File: StateFB.py
"""
This code shows a trajectory tracking of a chai  
of integrators system 

Note : To run the script type in terminal:
        pip install numpy
        pip install scipy
"""


import numpy as np
import matplotlib.pyplot as plt
from scipy import signal

# --- System matrices ---
A = np.array([[0, 1, 0, 0],
              [0, 0, 1, 0],
              [0, 0, 0, 1],
              [0, 0, 0, 0]])

B = np.array([[0],
              [0],
              [0],
              [1]])

B1 = np.array([0, 0, 0, 9.8])

# --- Desired poles ---
p = np.array([-1, -2, -3, -4])

# --- Pole placement (equivalent to MATLAB place) ---
K = signal.place_poles(A, B, p).gain_matrix  # shape (1,4)

# --- Simulation parameters ---
dt = 0.001
t_final = 50
s = np.arange(0, t_final + dt, dt)

# --- State variables ---
x = np.zeros((4, len(s)))
x[:, 0] = np.array([0, 0, 0, 0])

e = np.zeros((4, len(s)))
r = np.zeros((4, len(s)))
u = np.zeros(len(s))

# --- Simulation loop ---
for i in range(len(s) - 1):
    t = s[i]

    # Reference
    r[:, i] = np.array([
        4 + np.sin(0.5 * t),
        0.5 * np.cos(0.5 * t),
        -0.25 * np.sin(0.5 * t),
        -0.125 * np.cos(0.5 * t)
    ])

    # 4th derivative reference
    xpppp = 0.0625 * np.sin(0.5 * t)

    # Error
    e[:, i] = x[:, i] - r[:, i]

    # Control input (state feedback)
    u[i] = xpppp + (-K @ e[:, i]).item() / 9.8
    u[i] = u[i].item()  # convert from array to scalar

    # System dynamics (Euler integration)
    x_dot = A @ x[:, i] + B1 * u[i] - np.array([0, 0, 0, xpppp])
    x[:, i + 1] = x[:, i] + dt * x_dot

# --- Plot 1: Position ---
plt.figure()
plt.plot(s, x[0, :], label='x1')
plt.plot(s, r[0, :], label='x1_ref')
plt.xlabel("Time (s)")
plt.ylabel("Position (m)")
plt.legend()
plt.grid()

# --- Plot 2: Errors ---
plt.figure()
for i in range(4):
    plt.plot(s, e[i, :], label=f'e{i+1}')
plt.xlabel("Time")
plt.ylabel("Errors")
plt.legend()
plt.grid()

# --- Plot 3: Control input ---
plt.figure()
plt.plot(s, u, label='u')
plt.xlabel("Time")
plt.ylabel("Control input")
plt.grid()

plt.show()